plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2370,142 +2338,6 @@ bool producer_plugin_impl::process_unapplied_trxs(const fc::time_point& deadline
-   const auto         scheduled_trxs_size = sch_idx.size();
-   auto               sch_itr             = sch_idx.begin();
-   while (sch_itr != sch_idx.end()) {
-      if (sch_itr->delay_until > pending_block_time)
-         break; // not scheduled yet
-      if (exhausted || deadline <= fc::time_point::now()) {
-         exhausted = true;
-         break;
-      }
-      if (sch_itr->published >= pending_block_time) {
-         ++sch_itr;
-         continue; // do not allow schedule and execute in same block
-      }
-
-      if (blacklist_by_id.find(sch_itr->trx_id) != blacklist_by_id.end()) {
-         ++sch_itr;
-         continue;
-      }
-
-      const transaction_id_type trx_id         = sch_itr->trx_id; // make copy since reference could be invalidated
-      const auto                sch_expiration = sch_itr->expiration;
-      auto                      sch_itr_next   = sch_itr; // save off next since sch_itr may be invalidated by loop
-      ++sch_itr_next;
-      const auto next_delay_until = sch_itr_next != sch_idx.end() ? sch_itr_next->delay_until : sch_itr->delay_until;
-      const auto next_id          = sch_itr_next != sch_idx.end() ? sch_itr_next->id : sch_itr->id;
-
-      num_processed++;
-
-      // configurable ratio of incoming txns vs deferred txns
-      while (incoming_trx_weight >= 1.0 && itr != end) {
-         if (deadline <= fc::time_point::now()) {
-            exhausted = true;
-            break;
-         }
-
-         incoming_trx_weight -= 1.0;
-
-         auto trx_meta = itr->trx_meta;
-         bool api_trx  = itr->trx_type == trx_enum_type::incoming_api;
-
-         auto trx_tracker = _time_tracker.start_trx(trx_meta->is_transient());
-         push_result pr = push_transaction(deadline, trx_meta, api_trx, itr->return_failure_trace, trx_tracker, itr->next);
-
-         exhausted = pr.block_exhausted;
-         if (pr.trx_exhausted) {
-            ++itr; // leave in incoming
-         } else {
-            itr = _unapplied_transactions.erase(itr);
-         }
-
-         if (exhausted)
-            break;
-      }
-
-      if (exhausted || deadline <= fc::time_point::now()) {
-         exhausted = true;
-         break;
-      }
-
-      auto get_first_authorizer = [&](const transaction_trace_ptr& trace) {
-         for (const auto& a : trace->action_traces) {
-            for (const auto& u : a.act.authorization)
-               return u.actor;
-         }
-         return account_name();
-      };
-
-      try {
-         auto             start        = fc::time_point::now();
-         auto             trx_tracker  = _time_tracker.start_trx(false, start); // delayed transaction cannot be transient
-         fc::microseconds max_trx_time = fc::milliseconds(_max_transaction_time_ms.load());
-         if (max_trx_time.count() < 0)
-            max_trx_time = fc::microseconds::maximum();
-
-         auto trace = chain.push_scheduled_transaction(trx_id, deadline, max_trx_time, 0, false);
```

Comment:
Unfortunately, we cannot completely get rid of calling `push_scheduled_transaction` yet. We still need to call it for expired transactions in `generated_transaction_multi_index` so that we can expire any ones still existing. We may even need to do that over several blocks if there is a large queue to work through.

But we can greatly simplify the logic in this function because we just need to go through the index `by_expiration` and pick out the ones that have expired to call `push_scheduled_transaction` with them. Additionally, we can provide a small hard-coded number of them to process at the beginning of the block before giving other transaction types their turn.

Reply:
I would limit it to say a hardcoded 50ms deadline.

Reply:
I wonder if 50ms too much? Expiring a deferred transaction is pretty quick. So an attacker would need to schedule a lot of deferred transactions per block to continually use up the full 50 ms capacity per block. This would cost them RAM but only for 1 block which they could reuse in the next block. So really it only costs them the CPU used to schedule the smallest deferred transactions they can as quickly as possible.

Reply:
Sure 25ms or 10ms would work. I don't think we need it configurable.

Reply:
Oha, I missed this comment. Addressed it now. Make it to 10ms to be on safe side.